### PR TITLE
embeddable + criteria

### PR DESCRIPTION
**Pull request by EliosPride**

### CODE REVIEW

File: src/main/java/ma/hibernate/embeddable/Main.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                "content",
+                new PostMetadata(14L, "hashcode"));
+        PostService postService = new PostServiceImpl();
+        postService.add(first);
+
+        NewsPost second = new NewsPost(LocalDate.of(2006, 11, 2),
+                "murder",
+                "in Castlevania",
+                new PostMetadata(12L, "hashcode1"));
+        PostService postService1 = new PostServiceImpl();
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/Main.java]: @EliosPride Why have you created another postService?

File: src/main/java/ma/hibernate/embeddable/Main.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        postService1.add(second);
+
+        NewsPost third = new NewsPost(LocalDate.of(2012, 2, 2),
+                "Anton",
+                "eating soap",
+                new PostMetadata(2L, "hashcode2"));
+        PostService postService2 = new PostServiceImpl();
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/Main.java]: @EliosPride and again, and all places below

File: src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java [hunk 3]
```diff
@@ -11,11 +22,40 @@ public NewsPostDaoImpl(SessionFactory sessionFactory) {
 
     @Override
     public NewsPost save(NewsPost post) {
-        return null;
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java]: @EliosPride You use sessionFactory just 1 time, so I don't see any reason to save it to var

File: src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java [hunk 4]
```diff
@@ -11,11 +22,40 @@ public NewsPostDaoImpl(SessionFactory sessionFactory) {
...
+                transaction.rollback();
+            }
+            throw new DataProcessingException("can't save post to db", e);
+        }
     }
 
     @Override
     public List<NewsPost> getWithMetadataSizeLargerThan(long size) {
-        return null;
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java]: @EliosPride the same, and check also all other places

File: src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java [hunk 5]
```diff
@@ -11,11 +22,40 @@ public NewsPostDaoImpl(SessionFactory sessionFactory) {
...
+        try (Session session = sessionFactory.openSession()) {
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<NewsPost> criteriaQuery = criteriaBuilder.createQuery(NewsPost.class);
+            Root<NewsPost> root = criteriaQuery.from(NewsPost.class);
+            criteriaQuery.select(root);
+            Query<NewsPost> query = session.createQuery(criteriaQuery);
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java]: @EliosPride do it in chain (like stream API), no need to create so much variables for each method.

File: src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java [hunk 6]
```diff
@@ -11,11 +22,40 @@ public NewsPostDaoImpl(SessionFactory sessionFactory) {
...
+            List<NewsPost> notSortedList = query.getResultList();
+            List<NewsPost> postList = new ArrayList<>();
+            for (NewsPost element : notSortedList) {
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java]: @EliosPride use stream API

File: src/main/java/ma/hibernate/embeddable/factory/NewsPostDaoFactory.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    private static NewsPostDao newsPostDao;
+
+    protected NewsPostDaoFactory(SessionFactory sessionFactory) {
+        super(sessionFactory);
+    }
+
+    public static NewsPostDao getInstance() {
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+        if (newsPostDao == null) {
+            newsPostDao = new NewsPostDaoImpl(sessionFactory);
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/factory/NewsPostDaoFactory.java]: @EliosPride new NewsPostDaoImpl(HibernateUtil.getSessionFactory())

File: src/main/java/ma/hibernate/embeddable/model/PostMetadata.java [hunk 8]
```diff
@@ -1,9 +1,28 @@
...
+import javax.persistence.AttributeOverrides;
+import javax.persistence.Column;
+import javax.persistence.Embeddable;
+import java.util.Objects;
+
+@Embeddable
+@AttributeOverrides({
+        @AttributeOverride(name = "size", column = @Column(name = "data_size")),
+        @AttributeOverride(name = "hashSum", column = @Column(name = "data_hash_sum"))
+})
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/model/PostMetadata.java]: @EliosPride Why do you need this?

File: src/main/java/ma/hibernate/embeddable/model/NewsPost.java [hunk 9]
```diff
@@ -1,13 +1,44 @@
 package ma.hibernate.embeddable.model;
 
+import javax.persistence.*;
 import java.time.LocalDate;
+import java.util.Objects;
 
+@Entity
```
Comment (Garret-vin) [inline src/main/java/ma/hibernate/embeddable/model/NewsPost.java:7]: @EliosPride tableName?

File: src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java [hunk 10]
```diff
@@ -11,11 +22,40 @@ public NewsPostDaoImpl(SessionFactory sessionFactory) {
...
 
     @Override
     public List<NewsPost> getWithMetadataSizeLargerThan(long size) {
-        return null;
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+        try (Session session = sessionFactory.openSession()) {
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<NewsPost> criteriaQuery = criteriaBuilder.createQuery(NewsPost.class);
+            Root<NewsPost> root = criteriaQuery.from(NewsPost.class);
+            criteriaQuery.select(root);
```
Comment (al-pavlenko) [inline src/main/java/ma/hibernate/embeddable/dao/impl/NewsPostDaoImpl.java]: @EliosPride try this:
session.createCriteria(NewsPost.class)
.add(Property.forName("metadata.size").ge(size))
.list()

**Pull request merged**